src/main/sensors/gyroanalyse.c

```diff
@@ -1,35 +1,320 @@
+    for (int axis = 0; axis < 3; ++axis) {
+        fftResult[axis].centerFreq = 200; // any init value
+        biquadFilterInitLPF(&fftFreqFilter[axis], DYN_NOTCH_CHANGERATE, looptime);
+        biquadFilterInit(&fftGyroFilter[axis], FFT_BPF_HZ, 1000000 / FFT_SAMPLING_RATE, BIQUAD_Q, FILTER_BPF);
+    }
+}
 
-void gyroDataAnalyseInit(void)
-{
+// used in OSD
+const gyroFftData_t *gyroFftData(int axis) {
+    return &fftResult[axis];
 }
 
-void gyroDataAnalyse(const gyroDev_t *gyroDev, const gyro_t *gyro)
-{
-    UNUSED(gyroDev);
-    UNUSED(gyro);
+bool isDynamicFilterActive(void) {
+    return (IS_RC_MODE_ACTIVE(BOXDYNAMICFILTER) || feature(FEATURE_DYNAMIC_FILTER));
 }
 
-void gyroDataAnalyseUpdate(timeUs_t currentTimeUs)
-{
-    UNUSED(currentTimeUs);
+/*
+ * Collect gyro data, to be analysed in gyroDataAnalyseUpdate function
+ */
+void gyroDataAnalyse(const gyroDev_t *gyroDev, biquadFilter_t *notchFilterDyn) {
+    if (!isDynamicFilterActive()) {
+        return;
+    }
+
+    // if gyro sampling is > 1kHz, accumulate multiple samples
+    for (int axis = 0; axis < 3; ++axis) {
+        fftAcc[axis] += gyroDev->gyroADC[axis];
+    }
+    fftAccCount++;
+
+    // this runs at 1kHz
+    if (fftAccCount == fftSamplingScale) {
+        fftAccCount = 0;
+
+        //calculate mean value of accumulated samples
+        for (int axis = 0; axis < 3; ++axis) {
+            float sample = fftAcc[axis] / fftSamplingScale;
+            sample = biquadFilterApplyDF2(&fftGyroFilter[axis], sample);
+            gyroData[axis][fftIdx] = sample;
+            if (axis == 0)
+                DEBUG_SET(DEBUG_FFT, 2, lrintf(sample * gyroDev->scale));
+            fftAcc[axis] = 0;
+        }
+
+        fftIdx = (fftIdx + 1) % FFT_WINDOW_SIZE;
+    }
+
+    // calculate FFT and update filters
+    gyroDataAnalyseUpdate(notchFilterDyn);
+}
+
+void stage_rfft_f32(arm_rfft_fast_instance_f32 * S, float32_t * p, float32_t * pOut);
+void arm_cfft_radix8by2_f32( arm_cfft_instance_f32 * S, float32_t * p1);
+void arm_cfft_radix8by4_f32( arm_cfft_instance_f32 * S, float32_t * p1);
+void arm_radix8_butterfly_f32(float32_t * pSrc, uint16_t fftLen, const float32_t * pCoef, uint16_t twidCoefModifier);
+void arm_bitreversal_32(uint32_t * pSrc, const uint16_t bitRevLen, const uint16_t * pBitRevTable);
+
+typedef enum {
+    STEP_ARM_CFFT_F32,
+    STEP_BITREVERSAL,
+    STEP_STAGE_RFFT_F32,
+    STEP_ARM_CMPLX_MAG_F32,
+    STEP_CALC_FREQUENCIES,
+    STEP_UPDATE_FILTERS,
+    STEP_HANNING,
+    STEP_COUNT
+} UpdateStep_e;
+
+/*
+ * Analyse last gyro data from the last FFT_WINDOW_SIZE milliseconds
+ */
+void gyroDataAnalyseUpdate(biquadFilter_t *notchFilterDyn) {
+    static int axis = 0;
+    static int step = 0;
+    arm_cfft_instance_f32 * Sint = &(fftInstance.Sint);
+
+    uint32_t startTime = 0;
+    if (debugMode == (DEBUG_FFT_TIME))
+        startTime = micros();
+
+    DEBUG_SET(DEBUG_FFT_TIME, 0, step);
+    switch (step) {
```

Comment:
That's on purpose to do some steps at once and other on their own because of different run times.